packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +1009,92 @@ export class Select
+        if (!this.$fastController.isConnected) {
+            return;
+        }
 
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
+        const filteredOptions: ListOption[] = [];
+        let lastVisibleElement: Element | undefined;
+        for (const element of this.slottedOptions) {
+            let elementHidden = false;
+
+            if (element instanceof ListOptionGroup) {
+                if (element.hidden) {
+                    elementHidden = true;
+                    break; // no need to process hidden groups
+                }
+                const groupOptions = this.getGroupOptions(element);
+                const groupMatchesFilter = this.filterMatchesText(
+                    element.labelContent
                 );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
-        }
+                let allOptionsHidden = true;
+                groupOptions.forEach(option => {
+                    option.visuallyHidden = groupMatchesFilter
+                        ? false
+                        : this.isOptionHidden(option);
+                    if (!option.visuallyHidden) {
+                        filteredOptions.push(option);
+                    }
+                    allOptionsHidden = option.visuallyHidden && allOptionsHidden;
+                });
+                element.visuallyHidden = allOptionsHidden || element.hidden;
+                elementHidden = element.visuallyHidden;
+            } else if (isNimbleListOption(element)) {
+                element.visuallyHidden = this.isOptionHidden(element);
+                elementHidden = element.visuallyHidden;
+                if (!element.visuallyHidden) {
+                    filteredOptions.push(element);
+                }
+            }
 
-        this.options.forEach(o => {
-            if (isNimbleListOption(o)) {
-                if (!this.filteredOptions.includes(o)) {
-                    o.visuallyHidden = true;
-                } else {
-                    o.visuallyHidden = false;
+            if (!elementHidden) {
+                if (lastVisibleElement instanceof ListOptionGroup) {
+                    lastVisibleElement.showBottomSeparator = true;
+                }
+
+                if (element instanceof ListOptionGroup) {
+                    element.showTopSeparator = lastVisibleElement instanceof ListOption;
+                    element.showBottomSeparator = false;
                 }
             }
-        });
+
+            lastVisibleElement = elementHidden ? lastVisibleElement : element;
```

Comment:
This line can move into the `if (!elementHidden)` block and be simplified to `lastVisibleElement = element`.